### PR TITLE
Add flag to set CNI bin dir, and use it on gci nodes

### LINKED ISSUE #28563: `--network-plugin-dir` flag has ambiguous meanings
Ref http://kubernetes.io/docs/admin/network-plugins/

The `--network-plugin-dir` means two things under different settings of the `--network-plugin`:
- If `--network-plugin=cni`, the `--network-plugin-dir` is where the CNI config locates.
- If `--network-plugin=kubenet`, the `--network-plugin-dir` is where the plugin binaries can be found in addition to the default one (`/opt/cni/bin`).

IMO that having the flag to mix up the two related things makes me feel confused, can we split into two flags, one is for binary, one is for config files?
I understand that when using `--network-plugin=kubenet`, there's no need for the config dir, but I don't think that's sufficient we override the flag with another meaning...

Another benefits for having two flags is now with `--network-plugin=cni`, we will be able to specify the plugin binary search path other than `/opt/cni/bin/`.

cc @kubernetes/sig-network

### PR DESCRIPTION
**Pull request by bboreham**
**What this PR does / why we need it**:

When using `kube-up` on GCE, following #31023 which moved the workers from debian to gci, CNI just isn't working.  The root cause is basically as discussed in #28563: one flag (`--network-plugin-dir`) means two different things, and the `configure-helper` script uses it for the wrong purpose.

This PR adds a new flag `--cni-bin-dir`, then uses it to configure CNI as desired.

As discussed at #28563, I have also added a flag `--cni-conf-dir` so users can be explicit 

**Which issue this PR fixes** : fixes #28563

**Special notes for your reviewer**:

I left the old flag largely alone for backwards-compatibility, with the exception that I stop setting the default when CNI is in use.  The value of `"/usr/libexec/kubernetes/kubelet-plugins/net/exec/"` is unlikely to be what is wanted there.

**Release note**:

``` release-note
Added new kubelet flags `--cni-bin-dir` and `--cni-conf-dir` to specify where CNI files are located.
Fixed CNI configuration on GCI platform when using CNI.
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32151)

### CODE REVIEW

**Comment by bboreham** [comment]
Updated and rebased.  @freehan this is the PR I mentioned in #27027.

**Comment by bboreham** [comment]
Oh, one question: the default for NetworkPluginDir is in `pkg/apis/componentconfig/v1alpha1/defaults.go` but the defaults for CNI dirs are in `pkg/kubelet/network/cni/cni.go`.  Is it desirable to bring those into alignment?

File: cmd/kubelet/app/options/options.go [hunk 1]
```diff
@@ -147,7 +147,9 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.Int32Var(&s.LowDiskSpaceThresholdMB, "low-diskspace-threshold-mb", s.LowDiskSpaceThresholdMB, "The absolute free disk space, in MB, to maintain. When disk space falls below this threshold, new pods would be rejected. Default: 256")
 	fs.DurationVar(&s.VolumeStatsAggPeriod.Duration, "volume-stats-agg-period", s.VolumeStatsAggPeriod.Duration, "Specifies interval for kubelet to calculate and cache the volume disk usage for all pods and volumes.  To disable volume calculations, set to 0.  Default: '1m'")
 	fs.StringVar(&s.NetworkPluginName, "network-plugin", s.NetworkPluginName, "<Warning: Alpha feature> The name of the network plugin to be invoked for various events in kubelet/pod lifecycle")
-	fs.StringVar(&s.NetworkPluginDir, "network-plugin-dir", s.NetworkPluginDir, "<Warning: Alpha feature> The full path of the directory in which to search for network plugins")
+	fs.StringVar(&s.NetworkPluginDir, "network-plugin-dir", s.NetworkPluginDir, "<Warning: Alpha feature> The full path of the directory in which to search for network plugins or CNI config")
```
Comment (errordeveloper) [inline cmd/kubelet/app/options/options.go]: You probably want to mark it deprecated...

**Comment by errordeveloper** [comment]
I think you also need to run `make generated_files`, as you have updated `types.go`.

File: cmd/kubelet/app/options/options.go (hunk 1 above)
Comment (bboreham) [inline cmd/kubelet/app/options/options.go]: That would only be appropriate for the CNI usage of the flag; there is another meaning for non-CNI plugins.
Maybe non-CNI plugins should be deprecated, but that's another conversation.

**Comment by bboreham** [comment]
> I think you also need to run make generated_files, as you have updated types.go

OK, have run that and checked in the resulting changes.  Thanks!

File: cmd/kubelet/app/options/options.go [hunk 2]
```diff
@@ -147,7 +147,9 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
...
+	fs.StringVar(&s.CNIConfDir, "cni-conf-dir", s.CNIConfDir, "<Warning: Alpha feature> The full path of the directory in which to search for CNI config files")
```
Comment (caseydavenport) [inline cmd/kubelet/app/options/options.go]: Could we include the default values for these options?  Most of the other options have them listed.

Comment (bboreham) [inline cmd/kubelet/app/options/options.go]: Sure; any views on my question about which file the default should be implemented in?

**Comment by dcbw** [comment]
> Oh, one question: the default for NetworkPluginDir is in pkg/apis/componentconfig/v1alpha1/defaults.go but the defaults for CNI dirs are in pkg/kubelet/network/cni/cni.go. Is it desirable to bring those into alignment?

I feel like the defaults should be left up to each plugin type, either exec or CNI.  I think that was just an omission in Rajat's original CNI branch.  I don't think a kubelet global default is really appropriate here, since different plugins have different conventions.  So maybe remove the default from defaults.go and put it into exec.go?

**Comment by bboreham** [comment]
I added the default values to the usage messages and moved the 'exec' plugin default directory into `exec.go`.  Also squashed commits a bit.

> I think we can do CNI defaults in the Flags() calls

This doesn't seem to be the convention in `kubernetes/cmd/kubelet/app/options/options.go`.  All the flags there are using the current value from the struct (invariably the zero value) as a default, and those that display a default to the user have it hard-coded in the usage string.

**Comment by dcbw** [comment]
LGTM

**Comment by bboreham** [comment]
Could someone mark this as OK to test, please?  I'm quite keen to see CNI working on the default `kube-up` configuration.

**Comment by luxas** [comment]
ok to test

**Comment by caseydavenport** [comment]
@bboreham Thanks once again for doing the legwork for this! 

Can we please get this added to the v1.4 milestone, or get https://github.com/kubernetes/kubernetes/pull/31023 reverted?  This breaks CNI on GCE deployments...

**Comment by freehan** [comment]
Please fix the cni network plugin unit tests
```
FAIL	k8s.io/kubernetes/pkg/kubelet/network/cni [build failed]
```

Also check the error from `verify-flags-underscore.py`. Probably need to add new flag into `https://github.com/kubernetes/kubernetes/blob/master/hack/verify-flags/known-flags.txt`
```
FAILED   hack/make-rules/../../hack/verify-flags-underscore.py	3s
```

**Comment by thockin** [comment]
@pwittrock

**Comment by thockin** [comment]
Ugh.  Just to be clear, this is bad, but maybe not "force it into 1.4.0" bad.  I think this could wait for 1.4.1 to integrate.

**Comment by thockin** [comment]
I dislike how much of the individual net plugins are being exposed by flags now.  If we take a step back, could we define a single network-plugin flag that carried key-value params from flags to drivers?

**Comment by bboreham** [comment]
@thockin to be clear, you are reversing the position at https://github.com/kubernetes/kubernetes/issues/28563#issuecomment-230983339, where you wanted CNI in the name?

**Comment by freehan** [comment]
@bboreham do you need help to fix the tests?

**Comment by bboreham** [comment]
@freehan thanks for the offer; I think I am on top of it now.

**Comment by vishh** [comment]
I talked to @thockin offline and he is OK with this PR as-is. @freehan let's get this PR merged asap and have it cherry-picked into the 1.4 release branch.

**Comment by freehan** [comment]
Looks like the last missing piece:


```
FAILED   hack/make-rules/../../hack/verify-codecgen.sh	731s
```
Rerun hack/update-codecgen.sh

**Comment by freehan** [comment]
I think you have to run `hack/update-codecgen.sh` once again.

**Comment by freehan** [comment]
Thanks. LGTM. assuming tests will pass.

**Pull request merged**